2022/day7/day7.hs

```diff
@@ -0,0 +1,132 @@
+-- https://adventofcode.com/2022/day/7
+
+module Day7 where
+
+import Data.Function (on)
+import Data.List (find, intercalate, mapAccumL, minimumBy, sortBy)
+import Data.Map (fromListWith, toList)
+import Data.Maybe (fromMaybe)
+import System.Environment (getArgs)
+
+data DirectoryListing pathParts fileEntries = DirectoryListing [String] [([String], Int)]
+dirPath (DirectoryListing pathParts fileEntries) = pathParts
+files (DirectoryListing pathParts fileEntries) = fileEntries
+addFile ((DirectoryListing pathParts fileEntries), filename, size) =
+  DirectoryListing pathParts ((filename:pathParts, size):fileEntries)
+
+parentDir :: (DirectoryListing [String] [String]) -> (DirectoryListing [String] [String])
+parentDir (DirectoryListing (_:parent) fileEntries) = DirectoryListing parent fileEntries
+
+childDir :: (DirectoryListing [String] [String], String) -> (DirectoryListing [String] [String])
+childDir (DirectoryListing parent fileEntries, path) = DirectoryListing (path:parent) fileEntries
+
+rootDir :: (DirectoryListing [String] [String]) -> (DirectoryListing [String] [String])
+rootDir (DirectoryListing pathParts fileEntries) = DirectoryListing [] fileEntries
+
+formatPath :: [String] -> String
+formatPath pathParts = '/':(intercalate "/" (reverse pathParts))
+
+formatEntry :: (([String], Int) -> Maybe String) -> ([String], Int) -> String
+formatEntry indicator = (\(path, size) -> formatPath path ++ " : " ++ show size ++ fromMaybe "" (indicator (path, size)))
+
+formatEntries :: (([String], Int) -> Maybe String) -> [([String], Int)] -> String
+formatEntries indicator = (\entries -> intercalate "\n" (map (formatEntry indicator) entries))
+
+sortByPath :: [([String], size)] -> [([String], size)]
+sortByPath = sortBy (compare `on` (\(path, _) -> formatPath path))
+
+main = do
+  -- https://adventofcode.com/2022/day/7/input
+  commandLog <- readFile "input.log"
+
+  -- Process the log to build the list of directories with their total sizes
+  -- Flatten the ancestors to produce a full (flattened) tree of directories and sizes
+  let dirEntries = flattenAncestors (parentEntries (processLog commandLog))
```

Comment:
To my eyes, this line seems to do a lot.

It also seems like it might "read" better if it used Haskell's `&` operator a few times so that it read more like a data transformation process, like: `commandLog & processLog & parentEntries & flattenAncestors`.

Reply:
I hadn't yet learned the `&` operator; I'll try it out!